Fixes #13379 - Delete last_report when switching a host to build mode

roidelapluie opened the pull request.

dLobatog commented on test/unit/host_test.rb:
```diff
@@ -2255,12 +2255,13 @@ def to_managed!
     host = FactoryGirl.create(:host, :with_reports)
     assert host.reports.present?
     refute host.build?
+    refute host.last_report.nil?
```
I think you mean assert here, this test is failing

roidelapluie replied:
it looks like  FactoryGirl.create(:host, :with_reports) does not set last_report

dLobatog commented on test/unit/host_test.rb:
```diff
@@ -2255,12 +2255,13 @@ def to_managed!
     refute host.build?
+    refute host.last_report.nil?
     host.update_attributes(:build => true)
-    assert_empty host.reports.reload
```
Why remove this assert? It should still make sure the reports are deleted

roidelapluie replied:
mistake. I will fix it

dLobatog commented:
@roidelapluie The test you changed is failing, please take a look :smile: - Can you add the content on your bug description on Redmine to the commit description? Thanks!

dLobatog commented on test/unit/host_test.rb:
```diff
@@ -2255,12 +2255,14 @@ def to_managed!
     host = FactoryGirl.create(:host, :with_reports)
     assert host.reports.present?
     refute host.build?
+    refute host.last_report.blank?
```
This assertion is failing, there's no `last_report` in the host. I think it'd be nicer if you split this and made two different tests, one that tests the reports are deleted and other that tests last_report is cleared.

roidelapluie replied:
I think FactoryGirl.create(:host, :with_reports) should fill last_report. What do you think?

dLobatog replied:
That's true too, although I don't think `.create` should be necessary for testing this, `.build_stubbed` should suffice

Commit:
Fixes #13379 - Clear host.last_report when switching to build mode

Use case:

When a server with a Puppet client is rebuilt, foreman cleans the
reports but not the "last report" field.

When you switch a host from a Puppet-enablet hostgroup to a hostgroup
that does not install a Puppet client, the host is marked as "Out of
sync". Is should be marked as "No reports".

This would be consistent with the foreman behavior of deleting the
reports when switching to build mode.

roidelapluie commented:
@dLobatog updated

roidelapluie commented:
@dLobatog Build is green

dLobatog commented:
Merged as 6d10ed5a5e4b9fe71b847b6488c130baf542f7e1, thanks @roidelapluie!
